Feat bamboo emails

maslowalex opened the pull request.

maslowalex committed:
chore: mix deps.update

maslowalex committed:
feat: integrated Bamboo mailer

maslowalex commented on config/config.exs:
```diff
@@ -23,6 +23,29 @@ config :logger, :console,
+config :contact_form,
+  deliver_emails_to: "doma@doma.dev"
+
+config :contact_form, ContactForm.Mailer,
+  adapter: Bamboo.SMTPAdapter,
+  server: "smtp.domain",
+  hostname: "your.domain",
+  port: 1025,
+  username: {:system, "SMTP_USERNAME"},
+  password: {:system, "SMTP_PASSWORD"},
+  tls: :if_available, # can be `:always` or `:never`
+  allowed_tls_versions: [:"tlsv1", :"tlsv1.1", :"tlsv1.2"], # or {:system, "ALLOWED_TLS_VERSIONS"} w/ comma separated values (e.g. "tlsv1.1,tlsv1.2")
+  tls_log_level: :error,
+  tls_verify: :verify_peer, # optional, can be `:verify_peer` or `:verify_none`
+  tls_cacertfile: "/somewhere/on/disk", # optional, path to the ca truststore
+  tls_cacerts: "…", # optional, DER-encoded trusted certificates
+  tls_depth: 3, # optional, tls certificate chain depth
+  tls_verify_fun: {&:ssl_verify_hostname.verify_fun/3, check_hostname: "example.com"}, # optional, tls verification function
+  ssl: false, # can be `true`
+  retries: 1,
+  no_mx_lookups: false, # can be `true`
+  auth: :if_available # can be `:always`. If your smtp relay requires authentication set it to `:always`.
+
```
Important configurations, such as: `server`, `hostname`, and certificate details to be added.

cognivore replied:
Put this stuff into a secret, ignored config. Make an example of config like we do in do_auth.

Crash on boot if needed configuration settings aren't included. With an informative message, of course.

maslowalex committed:
fix: be explicit about when exactly the delivery should happen

cognivore commented on lib/contact_form_web/controllers/contact_controller.ex:
```diff
@@ -33,22 +35,11 @@ defmodule ContactFormWeb.ContactController do
-    File.mkdir("/tmp/messages")
-    File.write("/tmp/messages/#{file_name}", message)
-    File.mkdir("messages")
-    File.write("messages/#{file_name}", message)
```
Why remove file backend? I think it's ok to have.

cognivore commented on test/contact_form/notifiers/email_test.exs:
```diff
@@ -0,0 +1,53 @@
+defmodule ContactForm.Notifiers.EmailTest do
```
lgtm

cognivore reviewed (approved):
Overall OK, but needs an adjustment of how config laid out.

Also, it would be better to have a more informative commit message.

maslowalex committed:
chore: preserve original way of saving the messages into the filesystem

maslowalex committed:
feat: added requirement to have config/sensetive.exs file populated in order to be able to start project

maslowalex committed:
fix: remove grabage comments from sensetive.example.exs file

maslowalex committed:
fix: changed config naming to match doma conventions

maslowalex committed:
1. Get rid of ENV bullshit entirely for the time being.
2. Remove default values for phoenix bullshit (like salt and the private key seed) and put them into the secret.exs.example
3. Make sure to crash on lacking phoenix variables (like salt and the private key seed)
4. Use wildcard as in do_auth

cognivore commented on config/secret.exs.example:
```diff
@@ -0,0 +1,41 @@
+  no_mx_lookups: false,
+  auth: :if_available
+
+signing_salt = raise "Please set signing_salt"
```
WOW! THIS IS GENIUS!

cognivore replied:
I think that the same thing should go into "Mailer" config, right? Values that don't have defaults should be tagged with "raise".

cognivore reviewed (changes requested):
Very minor changes.

Please also explain when will `raise` be raised. At evaluation of the config so at compile time, right?

maslowalex commented:
> Very minor changes.
> 
> Please also explain when will `raise` be raised. At evaluation of the config so at compile time, right?

It seems like it raise during compilation:
```
➜  contact_form git:(feat-bamboo-emails) mix compile --force
** (RuntimeError) Please set signing_salt
    /Users/thunderbook/Work/doma/contact_form/config/dev.secret.exs:26: (file)
    /Users/thunderbook/Work/doma/contact_form/config/config.exs:12: (file)
    (stdlib 4.0.1) erl_eval.erl:748: :erl_eval.do_apply/7
    (stdlib 4.0.1) erl_eval.erl:492: :erl_eval.expr/6
    (stdlib 4.0.1) erl_eval.erl:136: :erl_eval.exprs/6
```

maslowalex committed:
raise on missing Mailer configuration

cognivore commented:
LGTM! Can't wait till we deploy it.
